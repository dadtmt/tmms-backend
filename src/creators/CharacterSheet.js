import R from 'ramda'
import React, { PropTypes } from 'react'
import { graphql } from 'react-apollo'
import { Alert, ListGroup, ListGroupItem, Panel } from 'react-bootstrap'
import {
  Field,
  FieldArray,
  propTypes as formPropTypes,
  reduxForm
} from 'redux-form'
import { branch } from 'recompose'

import { pageEditorId } from '../config'
import ButtonGroup from '../forms/components/ButtonGroup'
import CheckboxGroup from '../forms/components/CheckboxGroup'
import FieldGroup from '../forms/components/FieldGroup'
import ImageGroup from '../forms/components/ImageGroup'
import {
  CREATE_SHEET_MUTATION,
  UPDATE_SHEET_MUTATION
} from '../graphql/mutations'

const withCreateMutation = graphql(
  CREATE_SHEET_MUTATION,
  {
    props: ({ mutate }) => ({
      onSubmit: values => mutate({
        variables: {
          newSheet: {
            ...values,
            editorId: pageEditorId
          }
        }
      })
    })
  }
)

const withUpdateMutation = graphql(
  UPDATE_SHEET_MUTATION,
  {
    props: ({ mutate }) => ({
      onSubmit: values => mutate({
        variables: {
          sheet: R.omit('__typename')(values)
        }
      })
    })
  }
)

const renderCharacs = ({ fields }) => <div>
  <ButtonGroup
    glyph='plus'
    onClick={() => fields.push({})}
  />
  <ListGroup>
    {fields.map((charac, index) =>
      <ListGroupItem key={index} >
        <ButtonGroup
          glyph='trash'
          onClick={() => fields.remove(index)}
          title='remove'
        />
        <Field
          component={FieldGroup}
          controlId={`${charac}.name`}
          label='Characteristic:'
          name={`${charac}.name`}
          type='text'
        />
        <Field
          component={FieldGroup}
          controlId={`${charac}.value`}
          label='Value:'
          name={`${charac}.value`}
          type='number'
        />
      </ListGroupItem>
    )}
  </ListGroup>
</div>

renderCharacs.propTypes = {
  fields: PropTypes.object.isRequired
}

const renderGear = ({ fields }) => <div>
  <ButtonGroup
    glyph='plus'
    onClick={() => fields.push({})}
  />
  <ListGroup>
    {fields.map((item, index) =>
      <ListGroupItem key={index} >
        <ButtonGroup
          glyph='trash'
          onClick={() => fields.remove(index)}
          title='remove'
        />
        <Field
          component={FieldGroup}
          controlId={`${item}.name`}
          label='Name:'
          name={`${item}.name`}
          type='text'
        />
        <Field
          component={FieldGroup}
          controlId={`${item}.description`}
          label='Description:'
          name={`${item}.description`}
          type='text'
        />
        <Field
          component={ImageGroup}
          controlId={`${item}.image`}
          label='Image:'
          name={`${item}.image`}
        />
        <Field
          component={FieldGroup}
          controlId={`${item}.count`}
          label='Count:'
          name={`${item}.count`}
          type='number'
        />
      </ListGroupItem>
    )}
  </ListGroup>
</div>

renderGear.propTypes = {
  fields: PropTypes.object.isRequired
}

const CharacterSheet = ({ handleSubmit, pristine }) =>
<form onSubmit={handleSubmit}>
  <Field
    component={CheckboxGroup}
    label='player sheet'
    name='isPlayer'
  />
  <Field
    component={FieldGroup}
    controlId='name'
    label='Character name:'
    name='name'
    type='text'
  />
  <Field
    component={FieldGroup}
    controlId='description'
    label='description:'
    name='description'
    type='text'
  />
  <Field
    component={ImageGroup}
    controlId='image'
    label='image:'
    name='image'
  />
  <Panel header='Characteristics (add and edit)' >
    <FieldArray name='characs' component={renderCharacs} />
  </Panel>
  <Panel header='Gear (add and edit item)' >
    <FieldArray name='gear' component={renderGear} />
  </Panel>
  {!pristine && <Alert bsStyle='warning'>Modifications not saved</Alert>}
  <ButtonGroup disabled={pristine} glyph='save' title='save' type='submit' />
</form>

CharacterSheet.propTypes = {
  ...formPropTypes
}

const CharacterSheetForm = reduxForm({
  enableReinitialize: true,
  form: 'content'
})(CharacterSheet)

export default branch(
  R.pipe(
    R.path(['initialValues', 'id']),
    R.isNil
  ),
  withCreateMutation,
  withUpdateMutation
)(CharacterSheetForm)
