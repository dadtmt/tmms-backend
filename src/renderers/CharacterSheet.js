import React, { PropTypes } from 'react'
import { Label, Image, Panel, Table, Well } from 'react-bootstrap'

const CharacterSheet = ({
  characs,
  description,
  gear,
  image,
  isPlayer,
  name
}) => <Panel header={name}>
  {isPlayer && <Label>Player sheet</Label>}
  <Well>{description}</Well>
  {image && <Image src={image} rounded />}
  <Panel header='Characteristics'>
    <Table>
      <tbody>
        {characs.map(charac => <tr key={charac.name}>
          <td>{charac.name}:</td>
          <td>{charac.value}</td>
        </tr>)}
      </tbody>
    </Table>
  </Panel>
  <Panel header='Gear'>
    <Table>
      <tbody>
        {gear.map(item => <tr key={item.name}>
          <td>{item.name}</td>
          <td>{item.description}</td>
          <td>{item.image && <Image src={item.image} rounded width='50' />}</td>
          <td>{item.count}</td>
        </tr>)}
      </tbody>
    </Table>
  </Panel>
</Panel>

CharacterSheet.propTypes = {
  characs: PropTypes.array.isRequired,
  description: PropTypes.string.isRequired,
  gear: PropTypes.array.isRequired,
  image: PropTypes.string.isRequired,
  isPlayer: PropTypes.bool.isRequired,
  name: PropTypes.string.isRequired
}

export default CharacterSheet
