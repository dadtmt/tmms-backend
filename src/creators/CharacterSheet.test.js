import React from 'react'
import renderer from 'react-test-renderer'
import { createStore, combineReducers } from 'redux'
import { reducer } from 'redux-form'
import ApolloClient from 'apollo-client'
import { ApolloProvider } from 'react-apollo'

import CharacterSheet from './CharacterSheet'

const client = new ApolloClient()

const store = createStore(combineReducers({ form: reducer }))

it('renders correctly', () => {
  expect(
    renderer.create(<ApolloProvider client={client} store={store}>
      <CharacterSheet
        initialValues={{
          characs: [{
            name: 'Force',
            value: 4
          }],
          gear: [{
            count: 1,
            image: 'http://arc.png',
            name: 'Arc de feu'
          }],
          name: 'toto'
        }}
      />
    </ApolloProvider>).toJSON()
  ).toMatchSnapshot()
})
