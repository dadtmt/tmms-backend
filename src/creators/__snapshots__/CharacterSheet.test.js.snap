exports[`test renders correctly 1`] = `
<form
  onSubmit={[Function]}>
  <div
    className="checkbox"
    style={undefined}>
    <label>
      <input
        checked=""
        disabled={false}
        name="isPlayer"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        type="checkbox"
        value="" />
      player sheet
    </label>
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="name">
      Character name:
    </label>
    <input
      className="form-control"
      id="name"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder={undefined}
      type="text"
      value="toto" />
    <span
      className="help-block" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="description">
      description:
    </label>
    <input
      className="form-control"
      id="description"
      name="description"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder={undefined}
      type="text"
      value="" />
    <span
      className="help-block" />
  </div>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor={undefined}>
      image:
    </label>
    <input
      className="form-control"
      id={undefined}
      name="image"
      onBlur={[Function]}
      onChange={[Function]}
      onDragStart={[Function]}
      onDrop={[Function]}
      onFocus={[Function]}
      placeholder="Image url"
      type="text"
      value="" />
    <img
      className="img-rounded"
      src="" />
  </div>
  <div
    className="panel panel-default"
    id={undefined}>
    <div
      className="panel-heading">
      Characteristics (add and edit)
    </div>
    <div
      className="panel-body">
      <div>
        <div
          className="form-group">
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button">
            <span
              className="glyphicon glyphicon-plus" />
          </button>
        </div>
        <ul
          className="list-group">
          <li
            className="list-group-item">
            <div
              className="form-group">
              <button
                className="btn btn-default"
                disabled={false}
                onClick={[Function]}
                title="remove"
                type="button">
                <span
                  className="glyphicon glyphicon-trash" />
              </button>
            </div>
            <div
              className="form-group">
              <label
                className="control-label"
                htmlFor="characs[0].name">
                Characteristic:
              </label>
              <input
                className="form-control"
                id="characs[0].name"
                name="characs[0].name"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                placeholder={undefined}
                type="text"
                value="Force" />
              <span
                className="help-block" />
            </div>
            <div
              className="form-group">
              <label
                className="control-label"
                htmlFor="characs[0].value">
                Value:
              </label>
              <input
                className="form-control"
                id="characs[0].value"
                name="characs[0].value"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                placeholder={undefined}
                type="number"
                value={4} />
              <span
                className="help-block" />
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    className="panel panel-default"
    id={undefined}>
    <div
      className="panel-heading">
      Gear (add and edit item)
    </div>
    <div
      className="panel-body">
      <div>
        <div
          className="form-group">
          <button
            className="btn btn-default"
            disabled={false}
            onClick={[Function]}
            type="button">
            <span
              className="glyphicon glyphicon-plus" />
          </button>
        </div>
        <ul
          className="list-group">
          <li
            className="list-group-item">
            <div
              className="form-group">
              <button
                className="btn btn-default"
                disabled={false}
                onClick={[Function]}
                title="remove"
                type="button">
                <span
                  className="glyphicon glyphicon-trash" />
              </button>
            </div>
            <div
              className="form-group">
              <label
                className="control-label"
                htmlFor="gear[0].name">
                Name:
              </label>
              <input
                className="form-control"
                id="gear[0].name"
                name="gear[0].name"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                placeholder={undefined}
                type="text"
                value="Arc de feu" />
              <span
                className="help-block" />
            </div>
            <div
              className="form-group">
              <label
                className="control-label"
                htmlFor="gear[0].description">
                Description:
              </label>
              <input
                className="form-control"
                id="gear[0].description"
                name="gear[0].description"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                placeholder={undefined}
                type="text"
                value="" />
              <span
                className="help-block" />
            </div>
            <div
              className="form-group">
              <label
                className="control-label"
                htmlFor={undefined}>
                Image:
              </label>
              <input
                className="form-control"
                id={undefined}
                name="gear[0].image"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                placeholder="Image url"
                type="text"
                value="http://arc.png" />
              <img
                className="img-rounded"
                src="http://arc.png" />
            </div>
            <div
              className="form-group">
              <label
                className="control-label"
                htmlFor="gear[0].count">
                Count:
              </label>
              <input
                className="form-control"
                id="gear[0].count"
                name="gear[0].count"
                onBlur={[Function]}
                onChange={[Function]}
                onDragStart={[Function]}
                onDrop={[Function]}
                onFocus={[Function]}
                placeholder={undefined}
                type="number"
                value={1} />
              <span
                className="help-block" />
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
  <div
    className="form-group">
    <button
      className="btn btn-default"
      disabled={true}
      title="save"
      type="submit">
      <span
        className="glyphicon glyphicon-save" />
    </button>
  </div>
</form>
`;
